Storybook Composition - loading a prebuilt storybook into yours

ndelangen opened the pull request:
Storybook is used by teams, and it's actually quite common for multiple teams to work on their own isolated features, possibly in their own storybook. But it's also common these teams will share components between them.

In fact a very common scenario is that you're using some open source components, maybe a component library. They are using storybook, you are using storybook. But if you want to see their components -and the documentation that come with it- you have to go to their website/storybook or write stories for components that aren't yours.

This is wasting time. Storybook should have a mechanism to deal with this.

Presenting: storybook inception!

```js
// main.js
module.exports = {
  stories: [`${__dirname}/stories/*.*`],
  addons: [
    '@storybook/addon-roundtrip/register',
    '@storybook/addon-parameter/register',
    '@storybook/addon-preview-wrapper/register',
  ],
  refs: {
    inception: 'https://storybookjs-next.now.sh/dev-kits/iframe.html',
  },
};
```

The new `refs` lists all external (pre-build) storybooks you want storybook to load.

The storybook is loaded, and the stories defined in both will appear in the navigator.

Here's one of the challenges I faced: Stories might have the same name, and thus overwrite each other.
The default way storybook now deals with this is by prefixing all stories's kind with the id of the ref that loaded it.

There's a way to tweak this behaviour, because I expect the consumers will want some control over where included stories are placed. Here's how you do it:

```js
// manager.js
import { addons } from '@storybook/addons';

addons.setConfig({
  mapper: (refObject, storyObject) => {
    // return a storyObject, perhaps with a kind you customized
    return storyObject;
  },
});
```

If you return `null` then that story will be ignored, and not shown in the navigator.

---

The stories from 1 ref and another are completely isolated. Storybook will actually not merge them together, in fact storybook will load the actual iframes as specified in the ref, and 'juggle' multiple iframes and show the correct iframe for the currently selected story,

Currently this process isn't optimised, which means all refs are loaded from the start.
Once storybook has a way to pull the list of stories statically, we can lazy-load the iframe.


---

todo still

- [ ] expandable refs
- [ ] test auth flow
- [ ] documentation
- [ ] extract command with lazy puppeteer install
- [ ] preset to auto-load refs from dependencies

shilman commented on lib/api/src/index.tsx:
```diff
@@ -178,7 +185,46 @@ class ManagerProvider extends Component<Props, State> {
+          });
+        }
+      } else {
+        console.log('no refs loaded');
```
???

ndelangen replied:
Should be removed

shilman commented:
Hey @ndelangen, would love to add a more "convention over configuration" API. Let's discuss offline before merging. In the meantime will test out what you currently have!

bostanio commented on lib/api/src/index.tsx:
```diff
@@ -178,7 +185,44 @@ class ManagerProvider extends Component<Props, State> {
+            }
+          });
+        }
+      }
```
I find the code above super dense and it's kinda hard to read and grok what its doing. Can we break it up into a few named methods, add comments  etc. to make it more obvious what the logic is?

ndelangen replied:
I've tried to make it simpler, WDYT?

ndelangen commented:
@shilman let's talk about this in the near future.

ndelangen committed:
ADD refs to Sidebar

ndelangen committed:
Merge branch 'next' into core/inception

# Conflicts:
#	lib/api/src/index.tsx
#	lib/ui/src/components/sidebar/Sidebar.tsx

ndelangen committed:
IMPROVE types for Addons

ndelangen committed:
REORGANISE tools & preview

ndelangen committed:
FIX matching of source

ndelangen committed:
CHANGE ref to only refId

ndelangen committed:
ADD ability to emit event to target

ndelangen committed:
ADD ability to getData to get stories from refs

ndelangen committed:
ADD state for refs to be injected

ndelangen committed:
CLEANUP

ndelangen committed:
FIX zooming on refs

ndelangen committed:
CLEANUP

ndelangen committed:
FIX search in refs

ndelangen committed:
RENAME isInjected to initialState startInjected, state is preserved in preview itself

ndelangen committed:
REFACTOR preview so it renders many iframes

ndelangen committed:
CLEANUP

ndelangen committed:
CHANGE ManagerConsumer into a FunctionComponent that auto-memoizes & has MUCH better typings!

ndelangen committed:
REMOVE unneeded type annotations after improvement of types in lib/api

ndelangen committed:
Merge branch 'ts/better-api-consumer' into core/inception

# Conflicts:
#	lib/addons/package.json
#	lib/api/src/index.tsx
#	lib/ui/src/components/preview/preview.tsx
#	lib/ui/src/components/preview/toolbar.tsx
#	lib/ui/src/containers/preview.tsx

ndelangen committed:
Merge branch 'next' into core/inception

ndelangen committed:
CLEANUP

ndelangen commented on lib/addons/src/index.ts:
```diff
@@ -3,36 +3,21 @@ import global from 'global';
 import { ReactElement } from 'react';
```
I should cherry-pick this file change to `next`

ndelangen replied:
✔️

ndelangen commented on lib/api/src/init-provider-api.ts:
```diff
@@ -17,7 +20,12 @@ export interface Provider {
   renderPreview?: IframeRenderer;
   handleAPI(api: API): void;
-  getConfig(): Record<string, any>;
+  getConfig(): {
```
I could pull some of this out into a separate PR

ndelangen replied:
✔️

ndelangen commented on lib/api/src/lib/stories.ts:
```diff
@@ -2,40 +2,52 @@ import deprecate from 'util-deprecate';
   isLeaf: false;
   // MDX stories are "Group" type
-  parameters?: any;
+  parameters?: {
```
This TS improvement could be extracted as well

ndelangen replied:
✔️

ndelangen commented on lib/api/src/modules/layout.ts:
```diff
@@ -56,7 +56,7 @@ type PartialThemeVars = Partial<ThemeVars>;
 type PartialUI = Partial<UI>;
 
-interface Options extends ThemeVars {
+export interface UIOptions {
```
This isn't Inception specific either, extract to it's own PR

ndelangen replied:
✔️

ndelangen committed:
FIX getData throwing on undefined storyId

ndelangen committed:
CLEANUP

ndelangen committed:
Merge branch 'next' into core/ingestion

# Conflicts:
#	lib/addons/package.json
#	lib/api/src/init-provider-api.ts

ndelangen committed:
CLEANUP types of refs && RENAME ref.data to ref.stories && CHANGE a runtime ref config to be object instead of string

ndelangen committed:
CLEANUP

ndelangen committed:
ADD data.json fetch in node, for fetching data stories in before starting storybook in browser

ndelangen committed:
ADD example data for what data.json should look like

ndelangen committed:
FIX rename data to stories

ndelangen committed:
ADD data.json to cra-ts-kitchen-sink

ndelangen committed:
FIX default value of refs

ndelangen committed:
CLEANUP && FIX issue where component did not re-render when passing data THROUGH Consumer

ndelangen committed:
FIX bad depth field

ndelangen committed:
CLEANUP

ndelangen committed:
UPDATE data.json for example

ndelangen committed:
CHANGE to useState to keep track of which frames have been injected && ADD ready prop to ref so spinner goes away && transform when loading stories from data.json

ndelangen committed:
FIX version

ndelangen committed:
USE api.getData to retrieve storyObject

ndelangen committed:
CHANGE Root to also have refId optional

ndelangen committed:
ADD fullAPI to modules api initiaization

ndelangen committed:
CHANGE initProviderApi to be a void operation

ndelangen committed:
CHANGE addons module to use getData from fullAPI

ndelangen committed:
CHANGE jumpStory to use getData

ndelangen committed:
FIX addon-links

ndelangen committed:
Oops

ndelangen commented on lib/api/src/index.tsx:
```diff
@@ -197,19 +212,76 @@ class ManagerProvider extends Component<ManagerProviderProps, State> {
+          const refs = api.getRefs();
+          const [refId] = Object.entries(refs).find(([, { url }]) =>
+            `${url}/iframe.html`.match(source)
+          );
```
Seems shared logic

ndelangen committed:
CHANGE behavior of SidebarSearch, it should search all stories & refs

ndelangen committed:
FIX problem where when filtered, nodes could not be expanded

ndelangen committed:
IMPROVE filter code readability

ndelangen committed:
CHANGE search to only search in leafs, to prevent empty groups in the explorer

ndelangen committed:
CLEANUP

ndelangen committed:
FIX empty message

ndelangen committed:
ADD story for filtered dataset

ndelangen committed:
CLEANUP

ndelangen committed:
Merge branch 'next' into core/ingestion

ndelangen committed:
FIX tests

ndelangen committed:
FIX overflow of text & close button SearchField

ndelangen committed:
IMPROVE search fuzzyness

ndelangen committed:
CHANGE initial state of treeview expanded to all expanded when filtered, all contracted when unfiltered

ndelangen committed:
Merge branch 'next' into core/ingestion

# Conflicts:
#	lib/core/package.json

ndelangen committed:
CHANGE fetch to local version in TS

ndelangen committed:
FIX fetch

ndelangen committed:
FIX logo

ndelangen committed:
ADD title to example usage of refs

ndelangen committed:
ADD title to InceptionRef interface

ndelangen committed:
REMOVE content-background from <Main>

ndelangen committed:
IMPROVE types of setConfig

ndelangen committed:
CHANGE example of dark-theme tweak

ndelangen committed:
CHANGE 'empty because of filter' workings

ndelangen committed:
WIP

ndelangen committed:
RENAME loading to isLoading

ndelangen committed:
Never mutate in lib/api

ndelangen committed:
REFACTOR treeview & sidebar

ndelangen committed:
Merge branch 'next' into core/ingestion

# Conflicts:
#	lib/addons/package.json
#	lib/core/package.json

ndelangen committed:
MIGRATE addon-events

ndelangen committed:
CHANGE refs

ndelangen committed:
REMOVE warning

ndelangen committed:
OVERHAUL sidebar

ndelangen committed:
IMPROVE memoization of menu

ndelangen committed:
Revert "MIGRATE addon-events"

This reverts commit b227f774e4c0365823b3ca571f2a19d1ab1aa11e.

ndelangen committed:
ADD cors to netlify

ndelangen committed:
REMOVE loading of refs in node, load them in the browser always && ADD indicators to Sidebar

ndelangen committed:
IMPROVE typings of Icons

ndelangen committed:
FIX loading

ndelangen committed:
FIX ts error

ndelangen committed:
IMPROVE tooltips

ndelangen committed:
FIX bug of initial storyId of ref iframe

ndelangen committed:
REFACTOR api modules/stories for improved types

ndelangen committed:
IMPROVE readability of tooltip in darkmode

ndelangen committed:
IMPROVE text on tooltips for refs

ndelangen committed:
ADD SmoothHeight

ndelangen committed:
REMOVE SmoothHeight

ndelangen committed:
Make generated title look pretty

ndelangen committed:
Merge branch 'next' into core/ingestion

# Conflicts:
#	lib/addons/package.json
#	lib/core/package.json

ndelangen committed:
Merge branch 'next' into core/ingestion

ndelangen committed:
Merge branch 'next' into core/ingestion

ndelangen committed:
FIX loading appearance

ndelangen committed:
CHANGE listitem, cannot be loading && FIX selectedStory being rendered on top always

ndelangen committed:
Merge branch 'next' into core/ingestion

# Conflicts:
#	lib/api/src/lib/stories.ts
#	lib/api/src/modules/stories.ts

ndelangen committed:
Merge branch 'tech/use-getdata' into core/ingestion

# Conflicts:
#	examples/dev-kits/logo.svg
#	lib/components/src/tooltip/WithTooltip.tsx
#	lib/ui/src/app.tsx
#	lib/ui/src/containers/menu.tsx
#	lib/ui/src/containers/preview.tsx
#	lib/ui/src/containers/sidebar.tsx

ndelangen committed:
Merge branch 'next' into core/ingestion

ndelangen committed:
Merge branch 'tech/use-getdata' into core/ingestion

# Conflicts:
#	lib/ui/src/containers/preview.tsx

ndelangen committed:
FIX e2e by adding an ID back to the ListItem

REMOVE the expander if it's hidden anyway
REMOVE the attributes on <a> that shouldn't be there

ndelangen committed:
FIX package.json

ndelangen committed:
Merge branch 'next' into core/ingestion

ndelangen committed:
Merge branch 'next' into core/ingestion

ndelangen committed:
FIX ember docs generation if the component parameter is missing (for whatever reason)

ndelangen committed:
FIX unrenderable docs upon initialization

ndelangen committed:
CHANGE event target to full iframe id

ndelangen committed:
ADD key to li

ndelangen committed:
FIX typings

ndelangen committed:
CLEANUP

ndelangen committed:
Merge branch 'next' into core/ingestion

# Conflicts:
#	addons/docs/src/frameworks/ember/jsondoc.js
#	lib/components/src/tooltip/TooltipMessage.tsx
#	lib/core/src/client/preview/StoryRenderer.tsx
#	lib/ui/src/components/sidebar/ListItemIcon.tsx
#	netlify.toml

ndelangen committed:
CLEANUP

ndelangen committed:
CLEANUP mockdata, so it's not as huge

ndelangen committed:
FIX loader being invisible on light background or dark ones

ndelangen committed:
IMPROVE typings of toolbar

ndelangen committed:
FIX copy & eject functionality to work with refs

ndelangen committed:
REMOVE unneeded div

ndelangen committed:
CHANGE width of vertical dragger so it doesn't overlay the scrollbar

ndelangen committed:
ADD documentation for theme colors in dark & light

ndelangen committed:
CHANGE theme so when in dark mode lightest === darkest etc.

ndelangen committed:
IMPROVE theming on colors documentation page

ndelangen committed:
CHANGE Spaced to work across any type of children

ndelangen committed:
IMPROVE theming of sidebar && ADD stories for refs && IMPROVE how sidebar layout is done

ndelangen committed:
ADD authUrl to ref

ndelangen committed:
ADD gray Button

ndelangen committed:
CHANGE loader so Container its is re-usable

ndelangen committed:
REFACTOR Refs to match design

ndelangen committed:
IMPROVE stories for Refs

ndelangen committed:
IMPROVE the text when a user has authenticated

ndelangen committed:
IMPROVE typings of Button

ndelangen committed:
FIX deepscan issue

ndelangen committed:
REFACTOR Refs components, splitting them into single purpose components

ndelangen committed:
IMPROVE the error display in refs

ndelangen committed:
IMPROVE layout of long ref names & long version names

ndelangen committed:
CHANGE stories to share mockdata between them && ADD story for long names

ndelangen committed:
EXTRACTY RefHelpers & RefTypes to new module

ndelangen committed:
IMPROVE typings of menulist

ndelangen committed:
REMOVE need for shadow prop

ndelangen committed:
REMOVE unused file

ndelangen committed:
MIGRATE badge from css template literal to css objects

ndelangen committed:
no message

ndelangen committed:
Merge branch 'next' into core/ingestion

# Conflicts:
#	lib/ui/src/components/sidebar/Search.tsx

ndelangen committed:
Merge branch 'next' into core/ingestion

ndelangen committed:
CHANGE so all types & interfaces are exported

ndelangen committed:
FIX initial storyId being '*' in FramesRenderer

ndelangen committed:
ADD ability for MenuListItems to receive the listItem in the onClickHandler

Makes it possible to share a single callback for all Items
Makes for less re-renders
Makes for better performance

ndelangen committed:
FIX initial rendering of preview with "no story selected"

ndelangen committed:
ADD changeRefVersion API

ndelangen committed:
ADD onClick handler that refetches version to Ref VersionSelector

ndelangen committed:
ADD cors to local dev

ndelangen committed:
REMOVE really weird href = object type

ndelangen committed:
FIX missing key warning

ndelangen committed:
FIX state between switching versions && FIX wrong storyId being passed

ndelangen committed:
MOVE tooltips

ndelangen committed:
CHANGE so all interfaces & types are exported

ndelangen committed:
Merge branch 'tech/improvements-2' into core/ingestion

# Conflicts:
#	lib/components/src/tooltip/TooltipLinkList.tsx
#	lib/ui/src/components/sidebar/Tree/ListItem.tsx
#	lib/ui/src/components/sidebar/treeview/treeview.tsx

ndelangen committed:
REMOVE fluff from mockfiles

ndelangen committed:
Merge branch 'next' into core/ingestion

# Conflicts:
#	lib/ui/src/components/sidebar/Heading.stories.tsx
#	lib/ui/src/components/sidebar/Sidebar.stories.mdx
#	lib/ui/src/components/sidebar/Sidebar.stories.tsx
#	lib/ui/src/components/sidebar/Sidebar.tsx
#	lib/ui/src/components/sidebar/SidebarItem.stories.tsx
#	lib/ui/src/components/sidebar/SidebarStories.stories.tsx
#	lib/ui/src/components/sidebar/SidebarStories.tsx
#	lib/ui/src/components/sidebar/Tree/ListItem.tsx
#	lib/ui/src/components/sidebar/treeview/treeview.tsx
#	lib/ui/src/containers/sidebar.tsx

ndelangen committed:
FIX setRef call

ndelangen committed:
MOVE handlers into module's init

ndelangen committed:
FIX test

ndelangen committed:
Merge branch 'tech/improvements-3' into core/ingestion

# Conflicts:
#	lib/api/src/index.tsx
#	lib/api/src/modules/stories.ts
#	lib/api/src/modules/url.ts
#	lib/core/package.json

ndelangen committed:
FIX test

ndelangen committed:
Merge branch 'next' into core/ingestion

# Conflicts:
#	lib/api/src/modules/stories.ts

ndelangen committed:
REMOVE dummy data

ndelangen committed:
Merge branch 'next' into core/ingestion

# Conflicts:
#	lib/core/src/client/preview/start.ts

ndelangen committed:
Merge branch 'next' into core/ingestion

ndelangen committed:
Merge branch 'next' into core/ingestion

ndelangen committed:
Merge branch 'next' into core/ingestion

ndelangen committed:
ADD args to storyInput

ndelangen committed:
FIX missing args property on mockdata

ndelangen commented on lib/api/src/lib/stories.ts:
```diff
@@ -25,9 +28,11 @@ export interface Root {
 
 export interface Group {
   id: StoryId;
+  knownAs?: StoryId;
```
try to get rid of this and split the ref in the router

ndelangen committed:
REMOVE need for knownAs, since it can be deduced from refId + storyid

tmeasday commented on lib/api/src/modules/refs.ts:
```diff
@@ -0,0 +1,205 @@
+    },
+  };
+
+  const refs = Object.entries(api.getRefs());
```
```suggestion
  const refs = Object.entries(provider.getConfig().refs);
```

This is the point where we move the static refs into state.

tmeasday commented on lib/api/src/modules/refs.ts:
```diff
@@ -0,0 +1,205 @@
+      },
+    }),
+    {} as SubState['refs']
+  );
```
Do you think we should use [`lodash.mapValues`](https://lodash.com/docs/4.17.15#mapValues) for this kind of thing? I feel like it would be more readable and also probably more efficient. Or we could make our own version.

ndelangen replied:
turns out this is already guarded against in lib/core

tmeasday commented on lib/channel-postmessage/src/index.ts:
```diff
@@ -55,27 +55,43 @@ export class PostmsgTransport {
+      return new Promise((resolve, reject) => {
+        this.buffer.push({ event, resolve, reject });
+      });
+    }
```
Does this mean we buffer messages until the first frame appears, and then send them all to just that first frame?

We probably need to be careful that rather than the *first* frame, the **local** frame is the one we (a) wait for and (b) send the buffered messages to.

ndelangen replied:
I changed it so it's guaranteed that the local/main frame will get the flush.

tmeasday commented on lib/channel-postmessage/src/index.ts:
```diff
@@ -89,25 +105,56 @@ export class PostmsgTransport {
+      const list = nodes
+        .filter(e => {
+          try {
+            return !!e.contentWindow && e.dataset.isStorybook !== undefined && e.id === target;
```
Does this code work if `target` is undefined? Can you write a comment explaining what it is doing?

ndelangen replied:
if target is undefined, it will get the currentFrame:

> `return list.length ? list : this.getCurrentFrame();`

tmeasday commented on lib/channel-postmessage/src/index.ts:
```diff
@@ -89,25 +105,56 @@ export class PostmsgTransport {
+    return [];
+  }
+
+  private getCurrentFrame(): Window[] {
```
Seems odd that this returns an array.

ndelangen replied:
Seems quite likely to me, we'll have multiple active frames at a time in the future, I'll make this function name plural.

tmeasday commented on lib/channel-postmessage/src/index.ts:
```diff
@@ -89,25 +105,56 @@ export class PostmsgTransport {
+  private getCurrentFrame(): Window[] {
+    if (this.config.page === 'manager') {
+      const list: HTMLIFrameElement[] = [
+        ...document.querySelectorAll('#storybook-preview-wrapper span + iframe'),
```
I don't like this implementation detail from the `Preview` component leaking into the postmessage.

ndelangen replied:
agreed, will improve

tmeasday commented on lib/channel-postmessage/src/index.ts:
```diff
@@ -89,25 +105,56 @@ export class PostmsgTransport {
+      const list: HTMLIFrameElement[] = [
+        ...document.querySelectorAll('#storybook-preview-wrapper span + iframe'),
+      ];
+      return list.map(e => e.contentWindow);
```
What if this throws? (CORS)

ndelangen replied:
it can't, if accessing `.contentWindow` would throw, it has already thrown in the filter.

tmeasday commented on lib/channel-postmessage/src/index.ts:
```diff
@@ -117,6 +164,39 @@ export class PostmsgTransport {
+
+      return origin === event.origin;
+    });
+  }
```
It seems a bit weird that we will immediately fallback to the second method if *any* of the iframes don't allow CORS access. What about something more like the following:

```js
const [frame, ...remainder] = frames.filter(element => {
  try {
    return element.contentWindow === event.source;
  } catch (err) {
    // typical case, the iframe has not redirected at all
    return element.getAttribute('src') === event.source.document.location.href
  }
});

if (remainder.length > 0) throw new Error("something's weird");

if (frame) return [frame]

// We couldn't find the frame matching on window or current URL, see if there is a single frame with matching origin
const [frame, ...remainder] = frames.filter(e => new URL(element.getAttribute('src')).origin === event.origin);
if (frame && remainder.length === 0) return frame;

return null;
```

WDYT?

ndelangen replied:
I tweaked to WDYT of it now?

tmeasday replied:
It’s better but I still think we could check the `document.location.href` inside the iframe. Would that not work?

ndelangen replied:
that's not valid according to the types of the EventSource, which is why I didn't do it..

tmeasday replied:
I don't really get that? [MDN says](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) that `event.source` is:

> A reference to the window object that sent the message; you can use this to establish two-way communication between two windows with different origins.

So can't we use that to get a reference to the location? Haven't tried it obviously.

tmeasday commented on lib/core/src/server/manager/manager-webpack.config.js:
```diff
@@ -46,6 +48,17 @@ export default ({
+              });
+            `,
+          })
+        : null,
```
We might want to eventually generalize passing static data from `main.js` to the manager (and the preview for that matter), but this is 👍 for now.

ndelangen replied:
agreed

tmeasday commented on lib/ui/src/components/preview/FramesRenderer.tsx:
```diff
@@ -0,0 +1,82 @@
+      <Global styles={styles} />
+      {Object.entries(frames).map(([id, src]) => (
+        <Fragment key={id}>
+          {id === active ? <span key={`${id}-indicator`} /> : null}
```
As discussed I reckon we can probably just use a `useEffect` in `Iframe` to achieve this.

tmeasday reviewed (changes requested):
A few comments; I'd like to take another look after splitting manager-side storyIds in to `{ ref, storyId }`.

I didn't really review the `lib/ui` changes as they are pretty huge but it looks great!

ndelangen committed:
CHANGE to prefer getCurrentParameters

ndelangen committed:
REFACTOR to take current refId from router && REMOVE id mapping on ref stories

ndelangen committed:
REFACTOR to simplify mapping & adding refIds

ndelangen committed:
FIX test for css-resouces

ndelangen committed:
CHANGE setStories to be async && navigate to default story after state has been set && Do not navigate when viewmode is page or settings && FIX test for stories

ndelangen commented on lib/api/src/modules/refs.ts:
```diff
@@ -0,0 +1,205 @@
+}
+
+export type Mapper = (ref: InceptionRef, story: StoryInput) => StoryInput;
+export interface InceptionRef {
```
Yeah I'll rename it to ComposedRef, I din't want to name this Ref, because that might conflict with React.Ref in places

ndelangen committed:
RENAME InceptionRef to ComposedRef

ndelangen committed:
REFACTOR lib/api entrypoint & modules to not use default export

ndelangen committed:
FIX test

ndelangen commented on lib/api/src/index.tsx:
```diff
@@ -18,37 +18,26 @@ import {
+import * as layout from './modules/layout';
+import * as shortcuts from './modules/shortcuts';
+import * as url from './modules/url';
+import * as version from './modules/versions';
```
I think this nicely cleans up the file

ndelangen committed:
IMPROVE typings of lib/api

ndelangen committed:
Merge branch 'next' into core/ingestion

# Conflicts:
#	lib/api/src/index.tsx
#	lib/core/package.json

ndelangen committed:
REFACTOR so loading refs from config & loading refs from state are separate operations

ndelangen commented on lib/api/src/modules/refs.ts:
```diff
@@ -0,0 +1,205 @@
+      const { refs: fromState = {} } = store.getState();
+
+      return { ...fromConfig, ...fromState };
+    },
```
✅

ndelangen committed:
REMOVE unneeded code for ensuring a title exists on a ref

ndelangen committed:
CLEANUP

ndelangen committed:
IMPROVE getEventSourceUrl

ndelangen committed:
CLEANUP

ndelangen committed:
FIX type error

ndelangen committed:
FIX linting

ndelangen committed:
FIX tests

ndelangen committed:
CHANGE loader to always be round

ndelangen committed:
CLEANUP

ndelangen commented on lib/api/src/modules/refs.ts:
```diff
@@ -0,0 +1,205 @@
+
+      store.setState({
+        refs: {
+          ...(store.getState().refs || {}),
```
✅

ndelangen committed:
CLEANUP

ndelangen committed:
Merge branch 'next' into core/ingestion

ndelangen committed:
Merge branch 'core/ingestion' of github.com:storybookjs/storybook into core/ingestion

ndelangen committed:
CHANGE styling active frame from a funky span to a data attribute

ndelangen committed:
Merge branch 'tech/parallel-chromatic' into core/ingestion

# Conflicts:
#	lib/api/src/index.tsx
#	lib/core/package.json

ndelangen committed:
Merge branch 'next' into core/ingestion

ndelangen committed:
Merge branch 'next' into core/ingestion

ndelangen committed:
ADD an extract command to storybook cli

ndelangen committed:
ADD data.json to all examples automatically

ndelangen committed:
ADD data.json to all examples automatically

tmeasday commented on lib/cli/src/extract.js:
```diff
@@ -0,0 +1,88 @@
+import puppeteer from 'puppeteer';
```
Can we pull this into a separate PR?

ndelangen replied:
sure

ndelangen replied:
https://github.com/storybookjs/storybook/pull/10146

tmeasday commented on lib/router/src/utils.ts:
```diff
@@ -15,22 +15,25 @@ export const { toId, parseKind, sanitize, storyNameFromExport } = {
-const splitPathRegex = /\/([^/]+)\/([^/]+)?/;
+const splitPathRegex = /\/([^/]+)\/(?:(.*)_)?([^/]+)?/;
 
 export const parsePath: (path?: string) => StoryData = memoize(1000)(
```
Can we have a test for this?

ndelangen replied:
yes

ndelangen committed:
IMPROVE cli.extract && ADD missing dependency

ndelangen committed:
REMOVE support for setting options from refs

ndelangen committed:
IMPROVE debugging by adding colors to output of debug logs for channel && CHANGE channel event handler to only allow handling event from known sources

ndelangen committed:
CHANGE so frames for channel are filtered to only "ready" frames & flush events when frame loaded in after & on connect flush targets local/main frame

ndelangen committed:
CLEANUP

ndelangen committed:
CHANGE prefer getCurrentParameter over getParameters

ndelangen committed:
CHANGE name of getCurrentFrame to plural

ndelangen committed:
ADD tests for router utils

ndelangen committed:
ADD extract command to cli

ndelangen committed:
RENAME data.json to stories.json

ndelangen committed:
FIX tests

ndelangen committed:
Merge branch 'add/extract-command-to-cli' into core/ingestion

# Conflicts:
#	lib/cli/src/generate.js
#	scripts/build-storybooks.js

ndelangen committed:
ADD support for stories.json & metadata.json

ndelangen committed:
FIX linting error

ndelangen committed:
ADD puppeteer orb

ndelangen committed:
CHANGE to use puppeteer fix

ndelangen committed:
IMRPOVE puppeteer deps install

ndelangen committed:
CHANGE to use get-port

ndelangen committed:
Merge branch 'next' into core/ingestion

ndelangen committed:
Merge branch 'next' into add/extract-command-to-cli

ndelangen committed:
SIMPLIFY bugger logic

ndelangen committed:
FIX missing await

ndelangen committed:
FIX try using localhost, since it seems the server returned address doesn't seem to be accessible in CI

ndelangen committed:
FIX

ndelangen committed:
Merge branch 'add/extract-command-to-cli' into core/ingestion

ndelangen committed:
MOVE pretty logger to client-logger

ndelangen committed:
Merge branch 'next' into core/ingestion

# Conflicts:
#	lib/api/src/modules/layout.ts

ndelangen committed:
IMPROVE typing of client-logger

ndelangen committed:
CLEANUP

ndelangen committed:
CHANGE checkRef api to merge stories data & meta data

ndelangen committed:
Merge branch 'fix/generated-load-order' into core/ingestion

ndelangen committed:
Merge branch 'next' into core/ingestion

ndelangen committed:
Merge branch 'next' into core/ingestion

# Conflicts:
#	scripts/build-storybooks.js

ndelangen committed:
Merge branch 'next' into add/extract-command-to-cli

# Conflicts:
#	scripts/build-storybooks.js

ndelangen committed:
Merge branch 'next' into core/ingestion

# Conflicts:
#	examples/official-storybook/package.json
#	lib/channel-postmessage/package.json
#	lib/core/package.json

mrmckeb commented:
This looks really cool @ndelangen - it's a huge PR, but I get that a lot of this work wouldn't make sense in isolation.

I can already see a lot of use-cases for this, from versioning through to framework mixing, and of course importing Storybooks from other sources (such as consumed libraries).

@tmeasday, perhaps it's best if we get this in now, then create seperate PR(s) for authentication and UI updates? I think the sooner we get people playing with this, the better. I also feel like there will be quite a few tweaks and changes based on community feedback.

tmeasday commented:
> @tmeasday, perhaps it's best if we get this in now, then create seperate PR(s) for authentication and UI updates? I think the sooner we get people playing with this, the better. I also feel like there will be quite a few tweaks and changes based on community feedback.

Yeah, I'm down with that. There were a few outstanding things on the PR, I'm not sure if @ndelangen has addressed them yet? (It's such a beast it's hard to get a quick overview).

In any case it's minor stuff now.

mrmckeb commented:
It looked like most had been addressed when I was reviewing yesterday. Norbert walked me through it which helped. @ndelangen was there anything else?

ndelangen committed:
CHANGE to lazy-install puppeteer

ndelangen committed:
FIX use npx instead

ndelangen committed:
FIX ref to logger

ndelangen committed:
Merge branch 'next' into add/extract-command-to-cli

# Conflicts:
#	lib/cli/package.json

ndelangen committed:
Merge branch 'add/extract-command-to-cli' into core/ingestion

ndelangen committed:
CHANGE to use puppeteer-core install script

ndelangen committed:
Merge branch 'add/extract-command-to-cli' into core/ingestion

ndelangen committed:
FIX wait for script to finish

ndelangen committed:
FIX wait for script to finish

ndelangen committed:
FIX linting

ndelangen committed:
FIX path to install.js

ndelangen committed:
Merge branch 'add/extract-command-to-cli' into core/ingestion

# Conflicts:
#	lib/cli/src/extract.js
